File: 206.reverse-linked-list.cpp
/*
 * @lc app=leetcode id=206 lang=cpp
 *
 * [206] Reverse Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
   public:
    ListNode* reverseNode(ListNode* prev, ListNode* curr) {
        if (!prev || !curr) return prev;
        ListNode* next = curr->next;
        curr->next = prev;
        return reverseNode(curr, next);
    }

    ListNode* reverseList(ListNode* head) {
        if (!head) return head;
        ListNode* next = head->next;
        head->next = nullptr;
        return reverseNode(head, next);
    }
};
// @lc code=end

File: 206.reverse-linked-list.py
#
# @lc app=leetcode id=206 lang=python3
#
# [206] Reverse Linked List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    # recursion
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head

        reverse_next = self.reverseList(head.next)
        head.next.next = head
        head.next = None
        return reverse_next

    # iterative
#      def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
#         new_head = None

#         while head:
#             tail = head.next
#             head.next = new_head
#             new_head = head
#             head = tail
#         return new_head

    # tail recursion
    # def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
    #     def rec(head, acc):
    #         if head is None:
    #             return acc
    #         else:
    #             curr = head
    #             tail = head.next
    #             curr.next = acc
    #             return rec(tail, curr)
    #     return rec(head, None)
# @lc code=end
